Add remaining policy tests

joshsmith opened the pull request:
Closes #174 by adding the remaining policy tests not addressed in #320.
- Add category policy test
- Add role policy test
- Add skill policy test
- Add role skill policy test
- Add comment policy test
- Add preview policy test
- Add user policy test
- Add organization policy test
- Update abilities

JoshSmith commented:
@christopherstyles has it been provisioned for `code-corps-staging` only or also the production app?

JoshSmith commented:
👍 🐑 🇮🇹

christopherstyles commented:
@JoshSmith Oh that’s right – mind adding me to the staging app so I can set it up there as well?

JoshSmith commented:
@christopherstyles sure, did you want to add something else to this config then?

christopherstyles commented:
@JoshSmith No need to add anything else, so long as the `RAILS_ENV` on the staging app is set to `production`. It’ll use the same environment var.

JoshSmith commented:
@christopherstyles it is, just wasn't sure if it would be confusing which New Relic dashboard we're looking at (with naming). But probably doesn't matter all that much.

christopherstyles commented:
@JoshSmith Okay, gtg on staging now. Since there is no `staging` environment, it can’t be easily changed from the `newrelic.yml`. However, you should be able to change the name of the app from the “Account Settings” in the New Relic dashboard (you’ll need to, as only the app owner can do that).

JoshSmith commented:
@christopherstyles awesome, thanks! 👍

joshsmith committed:
Add remaining policy tests

Add category policy test

Add role policy test

Add skill policy test

Add role skill policy test

Add comment policy test

Add preview policy test

Update abilities

Add user policy test

Add organization policy test

Use assert and refute better

Use insert_pair

Add convenience fn for user and organization by role

Fix preview policy and preview changeset

sbatson5 commented on test/policies/category_policy_test.exs:
```diff
@@ -0,0 +1,29 @@
+  describe "create?" do
+    test "returns true when user is an admin" do
+      user = build(:user, admin: true)
+      assert create?(user) == true
```
Thoughts on just `assert create?(user)` as it looks for a truthy value?

sbatson5 commented on test/policies/category_policy_test.exs:
```diff
@@ -0,0 +1,29 @@
+
+    test "returns false if user is not an admin" do
+      user = build(:user, admin: false)
+      assert create?(user) == false
```
Thoughts on `refute create?(user)`?

sbatson5 replied:
There seem to be a few tests that explicitly check against true or false. I think we could leverage `assert` versus `refute`

sbatson5 commented on test/policies/comment_policy_test.exs:
```diff
@@ -0,0 +1,42 @@
+    end
+
+    test "returns false if someone else's record" do
+      user = insert(:user)
```
Thoughts on `[user, another_user] = insert_pair(:user)`?

sbatson5 commented on test/policies/organization_policy_test.exs:
```diff
@@ -0,0 +1,66 @@
+    test "returns false when user is pending member of organization" do
+      user = insert(:user)
+      organization = insert(:organization)
+      insert(:organization_membership, role: "pending", member: user, organization: organization)
```
We seem to have this block repeated in a few places where only the role varies. Thoughts on extracting this to a private method?

```
defp setup_user_organization_by_role(role) do
  ... all the codes but we pass in "pending"
end
```

This saves us a few lines but adds readability to the test so we can know exactly the intention of this block.

sbatson5 commented on web/policies/comment_policy.ex:
```diff
@@ -1,8 +1,9 @@
+  alias Ecto.Changeset
 
-  def create?(%User{}), do: true
+  def create?(%User{} = user, %Changeset{changes: %{user_id: creator_id}}), do: user.id == creator_id
```
I would expect a method name with a `?` to to imply it will return a bool. Thoughts on something like `can_create?`? This would probably be out of the context of this PR, but just a thought.

joshsmith replied:
@sbatson5 I think this is what canary expects.

sbatson5 commented on web/policies/organization_policy.ex:
```diff
@@ -8,6 +8,7 @@ defmodule CodeCorps.OrganizationPolicy do
   def create?(%User{admin: false}), do: false
 
+  def update?(%User{admin: true}, %Organization{}), do: true
   def update?(%User{} = user, %Organization{} = organization), do: user |> role_is_at_least_admin(organization)
```
Should there be a catch all?
`def update?(_, _), do: false`

joshsmith replied:
@sbatson5 I think it will return false if the role's not at least an admin.

sbatson5 replied:
I guess I wondering if there was ever a chance that we passed something that wasn't a `user` for the first param or `organization` for second. If we called `update?(nil, organization)`, we would get a clause failure.

joshsmith replied:
That should never happen as something else really major would have failed in that event (i.e. there's no current user even in the `conn` to be able to check against.

sbatson5 commented on web/policies/preview_policy.ex:
```diff
@@ -1,5 +1,7 @@
-  def create?(%User{} = _user), do: true
+  alias Ecto.Changeset
+
+  def create?(%User{} = user, %Changeset{changes: %{user: %{data: creator}}}), do: user.id == creator.id
```
What are we checking here? That the `creator` has changed in our changeset? Not sure I follow what this method is doing.

joshsmith replied:
The preview uses the `current_user` and the changeset assigns that user. We're checking that the creator of the preview is the user trying to create it.

sbatson5 reviewed (commented):
Just a couple of questions. Great to see more test coverage!  In general, it seems we can just leverage `assert` and `refute` for a lot of these tests. 👍

sbatson5 commented on test/policies/organization_policy_test.exs:
```diff
@@ -0,0 +1,58 @@
+
+    test "returns false when user is pending member of organization" do
+      [user, organization] = setup_user_organization_by_role("pending")
+      refute update?(user, organization)
```
👏 I think these tests read a lot better now! Very clear what we are testing.

joshsmith commented on web/helpers/authentication_helpers.ex:
```diff
@@ -51,15 +56,23 @@ defmodule CodeCorps.AuthenticationHelpers do
   end
   defp get_resource(_conn, model, :create), do: model.__struct__
 
   defp get_changeset_method(action), do: "#{action}_changeset" |> String.to_atom
```
@begedin this is really constraining and non-user friendly. In many cases we don't use `create_changeset` but just `changeset`, even for the `create`.

begedin replied:
@JoshSmith I was hoping this would establish a convention, much in the same way how canary already uses a convention to load a resource. However, we could extend the code to allow for an option to specify the method name.

begedin replied:
Also, a switch to `authy` would mean we would likely get rid of these and be more direct with which parameters we send into the various `can?` methods.

joshsmith commented on web/helpers/authentication_helpers.ex:
```diff
@@ -28,16 +28,21 @@ defmodule CodeCorps.AuthenticationHelpers do
-    conn |> assign(:changeset, changeset) |> authorize(changeset, action)
+    conn
+    |> assign(:changeset, changeset)
+    |> authorize(changeset, action)
```
This works really funky with `Preview` because we're doing an `assign_user` inside of the changeset. When this plug runs before the controller code itself is executed, we don't have the user that's going to be assigned in the controller yet inside the changeset that checks whether we're going to authorize it.

This is a bit 🐔 and 🍳 as a result.

begedin replied:
I'm not really sure how to resolve that, then. 

We could split up `load_and_authorize_changeset` into `load_changeset` and `authorize_changeset`. Canary already does the same with `_resource_`. For that specific action, since it's a `create`, we could load manually, using a locally defined plug method, then authorize using just `authorize_changeset`

joshsmith commented:
@begedin I'm pretty blocked on the above by `load_and_authorize_changeset`'s current functionality and have hit my limit on what I can accomplish tonight.

If you can give your thoughts on how we can refactor this to be less brittle, that would be really helpful.

joshsmith commented:
Going to temporarily remove the Preview policy changes from here and address again after we open a separate issue that makes preview changesets work like our others. The `assign_user` that we're doing there is unique and probably should not be.

joshsmith commented:
Also going to add an issue about adding `ja_resource` before proceeding.

joshsmith commented:
Latest change requires a change in Ember.

joshsmith commented:
@sbatson5 @begedin and anyone else interested, this should be ready for a re-review

begedin reviewed (approved):
Changes look good. Seems good to go to me 👍
